[RFC] Allow elastic agent to fail fast

shrikantnagori opened the pull request:
Summary: Today, on a segfault on a single trainer , we end up keeping the gpu on all ranks blocked for 5 minutes due to elastic agents barrier timeouts

Test Plan: Rely on existing test to validate . Looking to get some feedback on adding UTs

Differential Revision: D44929488

kiukchung commented:
thanks for the PR, could you elaborate on the failure scenarios (would help if you are specific about num workers and nodes and whether the node fails/exits on a signal, exception, or exits successfully) and the desired outcome behavior?

shrikantnagori commented:
>> thanks for the PR, could you elaborate on the failure scenarios (would help if you are specific about num workers and nodes and whether the node fails/exits on a signal, exception, or exits successfully) and the desired outcome behavior?

1. unsuccessful (non zero) exit code
2. hardware/host crashes

are the two common cases. Most often we run 8 gpu on a single Trainer Worker group(host) , so if training is using say 64 gpus. We will have 8 Trainer Worker groups, so if one fails, it waits 5 minutes for the other worker groups to finish for the above 2 failures.

kiukchung commented:
> > > thanks for the PR, could you elaborate on the failure scenarios (would help if you are specific about num workers and nodes and whether the node fails/exits on a signal, exception, or exits successfully) and the desired outcome behavior?
> 
> 1. unsuccessful (non zero) exit code
> 2. hardware/host crashes
> 
> are the two common cases. Most often we run 8 gpu on a single Trainer Worker group(host) , so if training is using say 64 gpus. We will have 8 Trainer Worker groups, so if one fails, it waits 5 minutes for the other worker groups to finish for the above 2 failures.

I see, so in terms of what behavior torchelastic currently exhibits, the worker failures can be bucketed into two categories:

1. Worker process exists with a non-zero exit code, either due to an catch-able exception or a signal 
2. Something outside the worker process exits with a non-zero exit code (e.g. torchelastic agent failure (doesn't happen all that often), host failure)

For 2) the exit barrier is irrelevant since the failure will be caught at the scheduler level (not torchelastic).

For 1) I see that the agent waits on the exit barrier (see [code](https://github.com/pytorch/pytorch/blame/master/torch/distributed/elastic/agent/server/api.py#L906)) on worker failures. And I wonder whether that is even needed... (cc @aivanou , @d4l3k  - do you guys remember why we added an exit barrier for worker failures?) On partial worker failures, the surviving workers on the other nodes will most likely get stuck until PTD times them out after `process_group.timeout` seconds or the node with the failed worker exits (after waiting on the exit barrier), whichever is shorter. In either case, this is a catastrophic failure with no reliable way to recover other than to restart all the workers, so there really isn't a great reason to wait on an exit barrier, but we might have added it purposefully based on an ops event in the past.

d4l3k commented:
@kiukchung I don't think I was involved with this before but if you restart one worker before all workers are restarting there's higher risk of getting bad overlapping rendezvous behavior where some workers are trying to re-rendezvous but others aren't and you may just hit a timeout in rdzv instead.

shrikantnagori commented:
I think removing the barrier is great as long as we can ensure there is something else that will mark the other workers as failed. Today timeout serves that purpose. So my take was to keep a small timeout to avoid changing the behavior on failure on other workers.

kiukchung commented:
> @kiukchung I don't think I was involved with this before but if you restart one worker before all workers are restarting there's higher risk of getting bad overlapping rendezvous behavior where some workers are trying to re-rendezvous but others aren't and you may just hit a timeout in rdzv instead.

Gotcha, FWIW in torchelastic, overlapping rdzv is avoided in this case (even without the exit barrier) since the elastic agent that observed the first worker failure would "re-announce" itself (as if it was re-added to the rdzv group). The monitoring loop on the other agents (runs every 30sec by default) will detect that there's a node waiting to be admitted and will SIGTERM all their respective workers and re-rdzv.

> I think removing the barrier is great as long as we can ensure there is something else that will mark the other workers as failed. 

Today we rely on PTD (the pg watchdog) to raise a `TimeoutError` naturally since if there are partial worker failures, then the collective op will get stuck, then eventually timeout.

> Today timeout serves that purpose. So my take was to keep a small timeout to avoid changing the behavior on failure on other workers.

If you are referring to the proces group `timeout`, then yes. If by `timeout` you were referring to the `exit_barrier()`, then I'm not 100% sure this was by design. The following is what is "supposed" to happen when there is a worker failure:

1. Assume `-j 2x4` (2 nodes, 4 workers on each node)
2. Assume rank 1 on node 0 failed
3. The elastic agent on node 0 will kill ranks 0, 1, 2, 3
4. If `max_restarts == 0`, then the elastic agent on node 0 will exit with a non-zero status, and hence the scheduler will deem the job as failed (it could restart the job according to the scheduler's job definition policy)
5. If `max_restarts > 0`, then the elastic agent on node 0 will "re-announce" itself to the rdzv group. The other workers in the group will notice that there's an agent waiting to "join" the group and will then proceed to kill all their local workers, and re-rdzv.

Now IIRC, this exit barrier was added due to the way PyPER was using torch.rpc and therefore had ranks that finish (either successfully or unsuccessfully) at different times.

kiukchung commented:
> can you point me to the process group timeout that you are referring to.

Take a look at https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group
![image](https://user-images.githubusercontent.com/43595115/232597514-0c21f90f-480a-4a3a-9520-f095076e6f7a.png)

> I can take the diff in a few paths from here

I think we should: 

1. Remove the `exit_barrier()` for the failure case (see [code](https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/agent/server/api.py#L906)).
2. Add a unittest testing for this behavior (run 2 agents, 2 workers each, with `max_retries=1` fail rank 1 the first time by design then make sure that the exit barrier is not hit and that the workers can re-rdzv fine without overlapping rdzv versions).

>  It looks like the exit barrier served no purpose other than rpc.shutdown in PyPer, but for that we have a timeout in the path added these days of 60 seconds, so that shall serve the same purpose.

If this is to get pyper jobs to fail-fast, can't we run those jobs with `exit_barrier_timeout=0`?

kurman reviewed (approved):
LGTM

kurman commented:
@kiukchung @d4l3k you are listed as code owners, will you be able to help with merging?

kiukchung commented on test/distributed/elastic/agent/server/test/local_elastic_agent_test.py:
```diff
@@ -963,6 +963,55 @@ def test_double_agent_fault_tolerance_etcd_v2(self):
         )
 
+    @patch("torch.distributed.elastic.utils.store.barrier")
+    def double_agent_clean_exit(self, barrier_mock):
```
(looks like my previous review comments have been lost for some reason here's the repeat).

This test isn't quite testing that the exit_barrier it not hit when there are worker failures for two reasons:

1. The test induces an agent (not a worker failure)
2. The mock is on the main (test) process but the agent is run on a sub-process. Therefore, the `barrier_mock` is trivially never called (python mocks don't push through subprocesses).

So this test needs to be re-written to induce a worker failure which can be done by writing an entrypoint function that raises an exception when being run from rank 0 and sleeps otherwise then replacing `conf.entrypoint._dist_sum` with this new function.

Rather than asserting that the patched `exit_barrier()` mock is not called, you'd have to assert that both agents finish with a non-zero exit code within at least `2 * monitor_interval` (in the test it is configured to be 0.1 sec)

kiukchung commented on test/distributed/elastic/agent/server/test/local_elastic_agent_test.py:
```diff
@@ -963,6 +1039,56 @@ def test_double_agent_fault_tolerance_etcd_v2(self):
+        # Validate that the processes finish close to each other.
+        # Using a slightly higher timeout than 2 * monitor_interval to make it less flaky
+        self.assertGreater(
+            1, time_between_processes, "Agents are not cleaned up until 2 * monitor_interval"
```
shouldn't this be a function of `monitor_interval`? something like:

```
self.assertGreater(2*monitor_interval, exit_interval_between_agents)
```

shrikantnagori replied:
So, i ran some tests and i did get numbers greater than 2 * monitor_interval with my code, monitor_interval is 0.01 , whereas i was getting values of 0.2-0.3 seconds exit_interval_between_agents . 
Also, there is no relevant code i found around 2* monitor_interval guarantee. 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/agent/server/api.py#L701

shrikantnagori replied:
cc @kiukchung to take a final look , i have used 2 * 0.1 as a check now, but the monitor_interval in the test is 0.01

kiukchung replied:
thanks, I'd say you plumb the `monitor_interval` in  `run_agent()` all the way through to the `get_worker_spec()` function so that you can set the monitor_interval to 0.1 explicitly without relying on the 0.01 default.

kiukchung commented on test/distributed/elastic/agent/server/test/local_elastic_agent_test.py:
```diff
@@ -963,6 +1039,56 @@ def test_double_agent_fault_tolerance_etcd_v2(self):
+        TEST_WITH_DEV_DBG_ASAN or TEST_WITH_TSAN,
+        "test incompatible with dev/dbg asan or tsan",
+    )
+    def test_double_agent_clean_exit_c10d(self):
```
the particular rdzv backend is irrelevant to this test so I'd rename this test to `test_no_exit_barrier_on_failure"

kiukchung reviewed (commented):
thanks for making the change, overall LGTM, have two small-ish comments

kiukchung commented on test/distributed/elastic/agent/server/test/local_elastic_agent_test.py:
```diff
@@ -963,6 +971,59 @@ def test_double_agent_fault_tolerance_etcd_v2(self):
+
+        # Validate that the processes finish close to each other.
+        # Using a slightly higher timeout than 2 * monitor_interval (0.01) to make it less flaky
+        monitor_interval = 0.1
```
you'd have to pass this `monitor_interval` as the agent config when you run the agent, or obtain the config that you ran the agent with and use the monitor_interval in the config object.

The way you have it here assumes that the agent was kicked off with `monitor_interval=0.1` which isn't quite true b/c the [`get_worker_spec()`](https://github.com/pytorch/pytorch/blob/main/test/distributed/elastic/agent/server/test/local_elastic_agent_test.py#L274) function defaults to `0.01` and `run_agent()` doesn't expose the `monitor_interval` flag (so you'll have to plumb it in)

shrikantnagori replied:
So, i plumbed the values. But the test remained flaky. The test needs more than 2 * monitor_interval for some values of monitor_interval, i stress tested with monitor_interval = 0.1 , 0.5 seconds and found that it did fail with 0.1  (1 failure out of 20 runs) , with 0.5 all my tests succeeded ( 20 stress tests), but i could not find logic that guaranteed shutdown within 2 * monitor_interval 

https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/agent/server/api.py#L701. Anyways, updated the diff with this suggestion and monitor_interval = 0.5 s

shrikantnagori replied:
cc @kiukchung  PTAL

kiukchung reviewed (approved):
LGTM, thanks for being so patient with all the review comments!

Commit:
[RFC] Allow elastic agent to fail fast (#99051)

Summary:
Today, on a segfault on a single trainer , we end up keeping the gpu on all ranks blocked for 5 minutes due to elastic agents barrier timeouts

Pull Request resolved: https://github.com/pytorch/pytorch/pull/99051

Test Plan: Rely on existing test to validate . Looking to get some feedback on adding UTs

Reviewed By: priyaramani, kurman

Differential Revision: D44929488

Pulled By: shrikantnagori

fbshipit-source-id: 44166bbcddc89f344be53265f98c8578fd14be89

kurman commented:
@pytorchbot merge

kurman commented:
release notes: distributed(elastic)

kurman commented:
@pytorchbot merge
